### DISCUSSION TITLE
Cras and Pulseaudio

### DISCUSSION THREAD

**Discussion by MeneerJansen** [dnschneid/crouton#4842] [General]
I recently had to re-install Linux on my Chromebook. I now have Debian 11 Bullsesye.

I cant remember how Cras (ChromeOS Audio Server) works in Crouton Linux. However, I cant start Pulseaudio apps like 'pavucontrol'. In my old Crouton choot I remember that I killed Cras this way:

`sudo initctl stop cras`

Doesn't work anymore. And '_alsamixer_' does not show my external USB audio device. Chrome browser in Linux does play audio but I used to use my Chromebook for music production. Trying to stop 'cras' with a 'systemctl' command throws me the error:

`Running in chroot, ignoring command 'stop'`

I cant use Jack Audio anymore (it sees NO sound device) which means that my Chromebook is not nothing more than a paperweight now. Help! 

Do I still need to kill cras to "do things audio"? And if so: how do I kill cras in Debian 11?

**Comment by TomTravis** [comment]
The Cras Device is a hardware driver for chomebooks  your maybe Mono 
I use xinit to start start pulseaudio ,, this is the bridge  that transfers rte audio to  Linux apps

.xinitrc is in your home dir 
editor (nano) .xinitrc
add
pulseaudio --start
exec xfce4-session 
^o to write is out 
^x to quit
at the prompt for cli
xinit

This should start the xorg server for graphics keyboard and pointer try some audio 

You should have audio controls in the menu

**Reply by MeneerJansen** [reply]
Thank you for your reply.

Starting PulseAudio is not the problem. I can do that via this good old command:

`pulseaudio --start -v`

The problem is that my external USB sound device ([Behringer 404](https://www.behringer.com/product.html?modelCode=P0BK1) which worked in the past ans still works on every other machine) does not appear in the tab "Configuration" in `pavucontrol`. Only a "device" called cras. This strikes me as odd because the real audio devices can be found with the command:

`cat /proc/asound/cards`

That actually does list the Behringer. It also lists the two true soundcards inside the Chromebook: HDMI audio and analog audio. But in no way can I use those! ??

Can anybody tell me if Crouton changed the way it handles audio and/or ALSA? I used to be able to use my external USB audio device just fine on this very same Chromebook.  

I want to stop the Cras service but neither of the following commands work:

- systemctl
- initctl
- service

When I try to use the command `service` to stop the cras service (SystemD) it says "service: command not found". Command `systemctl` says: "System has not been booted with systemd". But all those command do work on my Ubuntu Linux PC desktop computer. And systemd IS installed on the Debian chroot in Crouton.

Why can't I kill cras and simply use Alsa the way I'm used to in Lunux? Because it used to work **fine** in Crouton. ??

**Reply by TomTravis** [reply]
cras is the chrome book audio device linked to you HW audio chip

you an kill cras audio by turn is off in alza mixer
![Screenshot_20221029_164604](https://user-images.githubusercontent.com/46386261/198856566-59fa75e3-81c5-40bf-b907-8d35f432558c.png)
 is you want you then need some way to process and amp audio via digital USB out  my audio codec is mtk rt5650

**Comment by TomTravis** [comment]
Im using icewm audio via pulse audio you can turn off cras via a slider
![Screenshot_20221029_165747](https://user-images.githubusercontent.com/46386261/198856813-365f8329-e38e-492d-8be9-6eacfdc9e8f0.png)
you can see the pulse audio process is high on the chart

**Comment by TomTravis** [comment]
![Screenshot_20221029_172439](https://user-images.githubusercontent.com/46386261/198857293-476c513f-e204-444d-beb5-c55657f6842a.png)
audio is great nice and clean

**Comment by MeneerJansen** [comment]
Playing back audio on my Chr.b. is no problem. But I do **not** want to use the on-board audio card I want to use the one in the picture below. You connect such a device via an USB cable. That used to be possible, now it ain't.

![Behringer 404](https://i.postimg.cc/CLdt0Kgz/UMC404HD.jpg)

**Reply by TomTravis** [reply]
> Playing back audio on my Chr.b. is no problem. But I do **not** want to use the on-board audio card I want to use the one in the picture below. You connect such a device via an USB cable. That used to be possible, now it ain't.
> 
> ![Behringer 404](https://camo.githubusercontent.com/2d8cb58aa9a20cac756880a30ac478dbc5af6d8fbd8b60f7b9f31d61c66c64d5/68747470733a2f2f692e706f7374696d672e63632f434c6474304b677a2f554d4334303448442e6a7067)


driver install
https://packages.debian.org/bullseye/jackd2

**Comment by TomTravis** [comment]
if the device is known  the connect should be shown in the system 
The Install Shell and the config files for Crouton has no changed this I/F that i can tell

The chroot OS may have changed, it could be you enable the USB port and the device shows up
My camera(video/audio) and Mass storage work just fine USB

I believe this is device has I/O via usb your audio device

https://www.behringer.com/product.html?modelCode=P0BK1

interest read
https://linuxmusicians.com/viewtopic.php?t=17231
Your not alone it seem the USB port on the audio device has failed like others have
by Robert Luben October 30, 2022
Was great while it worked

I absolutely loved this thing, it was my daily driver for all audio in windows. Few months back we lost power briefly and it has not worked since. The mic inputs still work but it is no longer a USB device, tested on multiple PCs and its like it's not even plugged in. Tested with both USB power and wall power.

by Daniil October 16, 2022
pay your nerves for the price

This is a good card, I recorded several tracks on it, but there is one problem. Everytime when I stream or record video in OBS or NVIDIA Shadowplay (I've also tested several programms for recording). I can hear how sound disappears on miliseconds (with crackling sometimes). But for me it's ok and it happens only on my recordings... And this is really annoying. Reinstalled drivers, tried different USB ports, nothing helps.  Wanna change this card on something else because of this

**Comment by MeneerJansen** [comment]
The Linuxmusicians forum post you referred to is from _me_ in 2017.

Anyway I've powerwashed my Chr.b. and reverted to the factory ChromeOS version [link](https://chromeunboxed.com/fix-chromebook-bug-factory-reset-revert-powerwash). I appeared to have backuped my Ubunty Trusty chroot. Re-installed the back up and lo and behold it works again.

See screenshot of Alsamixer, nowadays those three audio devices do NOT turn up. Crouton used to be better! Wil never update Crouton or buy me a Chromebook.

![screeny](https://i.postimg.cc/tJv33dXy/Screenshot-from-2022-10-30-14-25-46.jpg)

So yes: in the old days we could access and use external USB hardware w/ Crouton, and now we can't.

Bye.

**Comment by MeneerJansen** [comment]
[yet another update]

Stop the press. When I run 'alsamixer' as root then it works! As a matter of fact I have to "do everything audio" as root (?!). Weird, because I'm member of the "audio" group. Can even run _jack_. Can use a Chromebook as a music production again. Phew!

**Reply by TomTravis** [reply]
well good enough jack is a lib it does not run is used as a driver to read and write and patch, 
the devices use it

try to keep the uses ID non root unless you have to i have some commands i had to with reinstall and reassign of Pulse audio and alsa mixer via chown The audio device should be seen as a sound card
yes your device is not new 2017 

On another subject icewm it working well now I had to purge reinstall and set chown to make is work audio parts

none of this was crouton Debootstrap installed maybe not the latest release
Jacks has been around for some time also groups file was updated for HWaudio and some others

**Comment by dnschneid** [comment]
crouton has two methods for audio.  One is via ChromeOS's audio server (cras), which it uses as an alsa device.  This plays nicely with ChromeOS audio and works for general use, but crouton's cras client can fall out of sync and need to be recompiled (via a crouton update) when there's a ChromeOS update.  Not relevant for your case.

The second way is to directly use the hardware interfaces, which is what you're doing.  Two things need to happen -- cras has to be stopped, and your chroot user needs to have the right permissions.

Generally you should be stopping cras from outside of the chroot; I wouldn't expect the initctl or service or whatever commands to work from inside, so it's surprising if that worked before.  It's always possible the interface to stopping cras has changed with a ChromeOS update, but it seems unlikely.

For permissions issues (which you seem to be having since running as root helps), crouton maintains a hwaudio group that matches the group id of the audio device outside the chroot.  You should be able to add your user to hwaudio and restart the chroot to get the permissions in order.  The hwaudio mapping gets updated every time you start the chroot, so crouton updates shouldn't affect anything.  It's possible a ChromeOS update has changed the group ownership of the audio devices, so crouton syncing the groups is no longer sufficient, so if your user is already in hwaudio but direct access doesn't work, that's probably the next thing to dig into.

**Reply by MeneerJansen** [reply]
> crouton has two methods for audio. One is via ChromeOS's audio server (cras), which it uses as an alsa device. This plays nicely with ChromeOS audio and works for general use, but crouton's cras client can fall out of sync and need to be recompiled (via a crouton update) when there's a ChromeOS update. Not relevant for your case.
> 
> The second way is to directly use the hardware interfaces, which is what you're doing. Two things need to happen -- cras has to be stopped, and your chroot user needs to have the right permissions.
> 
> Generally you should be stopping cras from outside of the chroot; I wouldn't expect the initctl or service or whatever commands to work from inside, so it's surprising if that worked before. It's always possible the interface to stopping cras has changed with a ChromeOS update, but it seems unlikely.

Hi there David!

Good to still see you pop up here every now and then! Thanks for making/maintaining Crouton. Crostini ain't no alternative for me since (I think) it will not let me access external USB hardware like a sound device yet (in the year 2022).

About stopping cras outside of the chroot: I take it that I should go into a terminal shell in ChromeOS? When I run `ps -e | grep cras` then I can see the cras process. As expected a simple kill command wont permanently stop it (it'll re-spawn). So I guess the I need some sort of Systemd command? But the commands `systemctl` and `service` doesn't work (not installed).  
 
> For permissions issues (which you seem to be having since running as root helps), crouton maintains a hwaudio group that matches the group id of the audio device outside the chroot. You should be able to add your user to hwaudio and restart the chroot to get the permissions in order.

I did the following and I might not completely understand what you mean (sorry). In the Debian Bullseye chroot I added myself to the group hwaudio. However, I'm the only member of that group so I don't think it existed before I added myself to it.

Am I doing it right? Because I still cannot acces my audio hardware as a regular user.

> The hwaudio mapping gets updated every time you start the chroot, so crouton updates shouldn't affect anything. It's possible a ChromeOS update has changed the group ownership of the audio devices, so crouton syncing the groups is no longer sufficient, so if your user is already in hwaudio but direct access doesn't work, that's probably the next thing to dig into.

Hmmm. Seems I have to dig into which group I must be a member of to have access to audio. There aren't that many groups (see list below) and I think that I'm a member of all the relevant ones (my username is michel)... 

root:x:0:
daemon:x:1:
bin:x:2:
sys:x:3:
adm:x:4:
tty:x:5:
disk:x:6:
lp:x:7:
mail:x:8:
news:x:9:
uucp:x:14:
man:x:12:
proxy:x:13:
kmem:x:15:
dialout:x:20:
fax:x:21:
voice:x:22:
cdrom:x:19:
floppy:x:11:
tape:x:26:
sudo:x:28:michel
audio:x:600:pulse,michel
dip:x:30:
www-data:x:33:
backup:x:34:
operator:x:37:
list:x:38:
irc:x:39:
src:x:40:
gnats:x:41:
shadow:x:42:
utmp:x:43:
video:x:27:michel
sasl:x:45:
plugdev:x:85:michel
staff:x:50:
games:x:60:
users:x:100:
nogroup:x:65534:
systemd-journal:x:101:
systemd-network:x:102:
systemd-resolve:x:103:
input:x:222:michel
kvm:x:105:
render:x:106:
crontab:x:107:
netdev:x:108:
hwaudio:x:18:michel
crouton:x:1001:michel
devbroker-access:x:403:
i2c:x:404:
serial:x:402:
wayland:x:999:michel
messagebus:x:104:
rtkit:x:109:
lightdm:x:110:
pulse:x:111:
pulse-access:x:112:
scanner:x:113:saned
avahi:x:114:
saned:x:115:
colord:x:116:
michel:x:1000:
ssh:x:117:

**Reply by dnschneid** [reply]
Does `sudo stop cras` work?  [The wiki](https://github.com/dnschneid/crouton/wiki/Access-audio-hardware-directly-(ALSA,-JACK)) probably has the latest steps from people actually using hardware audio.

It's expected that your user will be the only one in hwaudio.  The group number makes me think crouton is still keeping things in sync, so if that doesn't grant permissions, maybe there are bigger changes afoot.

**Reply by MeneerJansen** [reply]
> Does `sudo stop cras` work? [The wiki](https://github.com/dnschneid/crouton/wiki/Access-audio-hardware-directly-(ALSA,-JACK)) probably has the latest steps from people actually using hardware audio.
> 
> It's expected that your user will be the only one in hwaudio. The group number makes me think crouton is still keeping things in sync, so if that doesn't grant permissions, maybe there are bigger changes afoot.

That's remarkable, indeed `sudo stop cras` as well as `sudo initctl ` work in ChromeOS's shell. Just like they say in the Wiki (I completely forgot about the Wiki). These commands will, off course, not work in the (Debian Bullseye) chroot even though it says so in the Wiki: `initctl` is from Upstart which has been deprecated since about 2015. Maybe `initctl` works in ChromeOS's shell because my Chromebook/ChromeOS is insanely old (but nice!). I can't imagine that Google/ChromeOS still uses Upstart in 2022. 

I think the Wiki is from the time period of Ubuntu 14, which still used Upstart. They switched to _systemd_ with Ubuntu 15.

Anyway, I adding myself to the "hwaudio" group indeed did the trick. I can now run `alsamixer` and Jackaudio like it should. I apparently had to completely reboot the Chromebook and not just re-start the chroot (duh!). Silly me.

Thanks for the excellent info and tips! :)


P.S. Now for my `ssh` problems: a separate topic.

**Reply by dnschneid** [reply]
Glad that worked. You should update the wiki if it's out of date.

**Reply by MeneerJansen** [reply]
> Glad that worked. You should update the wiki if it's out of date.

Did just that ([link](https://github.com/dnschneid/crouton/wiki/Access-audio-hardware-directly-(ALSA,-JACK))).

**Comment by TomTravis** [comment]
David 

yes my setup for now is cras pulseaudio alsa-utils which includes now the alsamixergui kmix and some others 
this is the icewm start up pulse audio was exit for some reason
I had to purge  alsa* pulseaudio* and apt install new then now had kmix which manages the sounds for pulseaudio now

alsa mixer can change the sound device for use with pulseaudio which uses cras so jacks2 will help with USB sound devices have not tried BT or wifi yet

.xinitrc 
pulseaudio -k
pulseaudio -D --disallow-exit
kmix&
#exec mate-session
#exec startplasma-x11
exec icewm-session
